File: frontend/src/store/cartStore.ts
import { makeAutoObservable, computed } from 'mobx';
import type { CartItem, Product } from '../types';

class CartStore {
  items: CartItem[] = [];

  constructor() {
    makeAutoObservable(this, {
      total: computed,
      itemCount: computed,
    });
    this.loadFromStorage();
  }

  addItem(product: Product, quantity = 1) {
    const existing = this.items.find((item) => item.product.id === product.id);
    if (existing) {
      existing.quantity += quantity;
    } else {
      this.items.push({ product, quantity });
    }
    this.saveToStorage();
  }

  removeItem(productId: string) {
    this.items = this.items.filter((item) => item.product.id !== productId);
    this.saveToStorage();
  }

  updateQuantity(productId: string, quantity: number) {
    if (quantity <= 0) {
      this.removeItem(productId);
      return;
    }
    const item = this.items.find((i) => i.product.id === productId);
    if (item) {
      item.quantity = quantity;
      this.saveToStorage();
    }
  }

  clearCart() {
    this.items = [];
    this.saveToStorage();
  }

  get total() {
    return this.items.reduce((sum, item) => sum + item.product.price * item.quantity, 0);
  }

  get itemCount() {
    return this.items.reduce((sum, item) => sum + item.quantity, 0);
  }

  private saveToStorage() {
    localStorage.setItem('digital-store-cart', JSON.stringify(this.items));
  }

  private loadFromStorage() {
    try {
      const saved = localStorage.getItem('digital-store-cart');
      if (saved) {
        this.items = JSON.parse(saved) as CartItem[];
      }
    } catch {
      this.items = [];
    }
  }
}

export const cartStore = new CartStore();

File: backend/app/data/cart_store.py
from app.models.cart import CartItem, CartResponse
from app.data.database import get_product_by_id

# productId → quantity
_cart: dict[str, int] = {}


def _build_response() -> CartResponse:
    items: list[CartItem] = []
    for product_id, quantity in _cart.items():
        product = get_product_by_id(product_id)
        if product is not None:
            items.append(CartItem(product=product, quantity=quantity))

    total = round(sum(item.product.price * item.quantity for item in items), 2)
    return CartResponse(items=items, total=total)


def get_cart() -> CartResponse:
    return _build_response()


def add_item(product_id: str, quantity: int = 1) -> CartResponse | None:
    if get_product_by_id(product_id) is None:
        return None
    _cart[product_id] = _cart.get(product_id, 0) + quantity
    return _build_response()


def update_item(product_id: str, quantity: int) -> CartResponse | None:
    if product_id not in _cart:
        return None
    _cart[product_id] = quantity
    return _build_response()


def remove_item(product_id: str) -> CartResponse | None:
    if product_id not in _cart:
        return None
    del _cart[product_id]
    return _build_response()


def clear_cart() -> None:
    _cart.clear()
